### PR TITLE
Node script to copy projects from staging or prod

### LINKED ISSUE #1542: feat: Backup/Restore project as zip file
**Is your feature request related to a problem? Please describe.**
There is currently no simple way to download a Language Forge project for purposes of reproducing errors and fixing up projects.

**Describe the solution you'd like**
A menu option to download the project as a zipfile (export) and one to import the project as a zipfile, and the right processes in place to make these actions possible.

Data to be backed up

- [ ] mongodb project db (includes entries, comments and activity)
- [ ] mongodb project doc inside sf db
- [ ] assets folder
- [ ] s/r repo folder
- [ ] s/r state file
- [ ] user records involved in the project

UI should have an option to include filesystem data or leave it out of the backup (results in broken media links)

Backup behavior

- [ ] create zip file on server, then link to the file for download
- [ ] how to clean up zip file on server (/tmp file location?)...is this even necessary in an ephemeral container?

Restore behavior (dev command line use case)

- [ ] resolve dangling userRefs - create stub anon users for these userRefs; check project integrity?
- [ ] add user to project as owner/manager

Implementation notes

- have to install `mongo-tools` in the php app image in order to access command line tools e.g. `mongoexport` and `mongoimport`. `mongodump` and `mongorestore` are not suitable for this use case as we cannot easily choose which collections fields to export/import.

### PR DESCRIPTION
**Pull request by rmunn**
## Description

Bash script (on Windows, you'll want to run it with gitbash) to copy projects from staging or production to your local Docker dev environment.

Usage: First, edit the script and make sure the `staging_context` and `prod_context` values match the names you've given to your Kubernetes contexts. (If unsure, run `kubectl config get-contexts` to see what context names you have on your system).

Then run `node backup.mjs MongoID`, e.g. to copy https://staging.languageforge.org/app/lexicon/5dbf805650b51914727e06c4, you'd copy the Mongo ID out of that URL and run `backup.sh 5dbf805650b51914727e06c4`.

Alternately, you can just paste the URL as the command-line argument, at which point the script will automatically extract the project ID. Be careful to quote the URL, as some characters might have special meaning to the shell. (For example, on Linux, the `!` character means "Find a previous command that starts with this text". If you don't put quotes around the URL, you'll get an error from Bash saying `bash: !/editor/entry/5dbf806cbea602641cc27e61?sortBy=Default: event not found`. And they must be **single** quotes, because double-quotes don't remove the special meaning of `!`).

## Dependencies

You need `kubectl` and `docker` installed. Also, the script first tries to copy assets using `kubectl cp`, but if that fails, it falls back to `rsync`. On Windows, you might need to install [a Windows build of rsync from the msys project](https://packages.msys2.org/package/rsync).

## Checklist

- [X] I have labeled my PR with: bug, feature, engineering, security fix or testing
- [X] I have performed a self-review of my own code
- [X] I have reviewed the title & description of this PR which I will use as the squashed PR commit message
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have enabled auto-merge (optional)

## Testing

Testers, use the following instructions against our staging environment. Post your findings as a comment and include any meaningful screenshots, etc.

_Describe how to verify your changes and provide any necessary test data._

- Go to staging.languageforge.org and pick a project you'd like to copy to your local LF instance
- Copy its project ID
- Run `backup.sh copiedProjectID`
- Load localhost:8080 and make sure the project you just copied is available
- If backup.sh did not print any tar errors, then you may have all the project's sound and image assets
- If it *did* print tar errors, you may only have some but not all of the assets

### CODE REVIEW

**Comment by rmunn** [comment]
In our discussion from https://github.com/sillsdev/web-languageforge/issues/1542#issuecomment-2121843964 we talked about modifying all the userRef values from the lexicon (entries, comments, and so on). The script does not do that yet. So far I haven't encountered errors due to userRef values pointing to non-existent users, but I haven't tested that extensively. That might end up being unnecessary, but more testing is needed to prove that.

**Comment by hahn-kev** [comment]
I'm wondering if it would make more sense to write a script like this in JS which has a pretty high confidence of running cross platform, as it is Chris has stopped working on Mac so the majority of the team will be running this on windows.

As it is bash scripts are difficult to maintain since most of us don't write bash unless we need to. Additionally with JS we could just use kubectl to open a port to the db and use a mongo connection directly which would make it much simpler to write than a line like this `admin_id=$(docker exec lf-db mongosh -u admin -p pass --authenticationDatabase admin scriptureforge --eval "db.users.findOne({username: 'admin'}, {_id: 1})" | cut -d"'" -f 2)`
thoughts?

**Comment by megahirt** [comment]
@rmunn I really wanted to test this however my windows machine doesn't have kube contexts or a wireguard tunnel setup :(  I need my existing tunnel and contexts which are on my mac at home.  So maybe tomorrow I will test on windows.

**Comment by rmunn** [comment]
I'm willing to rewrite in JS if necessary, though I'd like to see how it performs in a Git Bash environment first, to avoid unnecessary work if it does turn out to be unnecessary.

**Comment by megahirt** [comment]
> I'm willing to rewrite in JS if necessary, though I'd like to see how it performs in a Git Bash environment first, to avoid unnecessary work if it does turn out to be unnecessary.

I asked chatgpt to port it over to Node Typescript and I thought it did a good first pass: https://chat.team-gpt.com/lt-lexical-tools/664585252bf6048a1b9a3f67

**Comment by rmunn** [comment]
It seems `kubectl exec` is just not reliable, so the mongodump/mongorestore approach to copying the project database is not going to work. I just added a loop to the Node.JS script to keep running mongodump/mongorestore until it succeeds, and my most recent run has now been going for over *two hours* without mongodump succeeding one single time. And that's on a project with just 30 entries, which takes just a few seconds to mongodump when it succeeds.

I'm going to rewrite the mongodump/mongorestore step to use a MongoClient connection instead. Since I can't use the `.copyDatabase()` feature that Mongo removed in version 4.2, I'll list the collections on the remote database, grab all the data from one collection at a time, and on the local database, drop the existing collection before doing an `insertMany` or `bulkWrite` operation to load the data.

File: backup.mjs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+console.log(`Local admin ID: ${adminId}`);
+console.warn("If that doesn't look right, hit Ctrl+C NOW");
+
+await portForwardingPromise;
+console.warn("Port forwarding is ready. Setting up remote Mongo connection...");
+
+const remoteConnStr = `mongodb://localhost:${remoteMongoPort}`;
+remoteConn = await MongoClient.connect(remoteConnStr);
+
+const remoteAdmin = await remoteConn.db("scriptureforge").collection("users").findOne({ username: "admin" });
```
Comment (megahirt) [inline backup.mjs:136]:
```suggestion
```
Probably remove this since it's not used?

Comment (rmunn) [reply backup.mjs]: Done in commit 6d9bd5561.

**Comment by rmunn** [comment]
NOTE: If you get an error like `tar: ./audio: File removed before we read it`, then it means you copied a project where one of the two directories (audio and pictures) was a broken symlink. This isn't likely to happen often, so I'm not adding mitigation to the script. The answer is simply to go into `deploy/app`, delete the broken symlink, and replace it with an empty directory.

If this happens a lot, I'll open a separate issue to track that bugfix.

File: backup.mjs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+// ===== EDIT THIS =====
+
+const stagingContext = "dallas-rke";
+const prodContext = "aws-rke";
+
+// Choose one, comment out the other
+// Alternately, pass a second argument to this script: "qa" or "staging" select staging, "live" or "prod" or "production" select production
+
+let contextName = "staging";
+let context = stagingContext;
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backup.mjs:19]: It's only overwritten if you pass a fourth argument, but I can change this to `defaultContext` and clean it up that way.

File: backup.mjs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+console.warn("Getting name of remote app pod...");
+const pod = run(
+  `kubectl --context="${context}" --namespace=languageforge get pod -o jsonpath="{.items[*]['metadata.name']}" -l app=app --field-selector "status.phase=Running"`,
+);
+console.warn("Trying to fetch assets tarball with kubectl cp...");
+let failed = false;
+try {
+  execSync(
+    `kubectl --context="${context}" --namespace=languageforge cp ${pod}:/tmp/assets-${dbname}.tar ${tempdir}/assets-${dbname}.tar`,
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backup.mjs:294]: Figuring out which assets are available to download is not something we have an API for, so although any one individual asset would be simpler to download via HTTP, downloading them all at once is not simple.

**Review by myieye — CHANGES_REQUESTED** [review]
This is like crazy white magic to me 🪄 🤓 👏 

It seems to work 🚀 , but not flawlessly just yet. I'm getting:
```
  throw new Error(`Unexpected result from readlink ${name}: ${result}`);
        ^

Error: Unexpected result from readlink /var/www/html/assets/lexicon/sf_test-chris-01/audio: -n no'
```

Might have something to do with windows.
Changing the occurences of `echo -n` to use printf (e.g.) `printf "no"` seems to fix it for me.

File: package-lock.json
Comment (myieye) [inline package-lock.json:1]: I've got a `pnpm-lock.yaml` file. Didn't we change to pnpm? I'm confused 🤔.

**Comment by rmunn** [comment]
> Might have something to do with windows. Changing the occurences of `echo -n` to use printf (e.g.) `printf "no"` seems to fix it for me.

Grrr. Windows, what are you doing? That `echo` was clearly inside quotes, it should not have been processed by your shell! It was supposed to be part of the `sh -c` input.

It's those little subtle differences that get you when trying to write cross-platform scripts. Another thing I could do here is give up on the `-n` option and instead strip newlines from the command output before comparing it to "yes" or "no". (The `-n` option to echo means "no newline"; normally echo will automatically add a newline after the text you give it). I think I'll do that, as `-n` might be a little too much magic for Windows.

File: package-lock.json
Comment (rmunn) [reply package-lock.json:1]: I think I branched this off before making that change. I'll rebase off of develop to fix the merge conflict and get the dependency correctly into pnpm-lock.json.

File: backup.mjs [hunk 4]
```diff
@@ -0,0 +1,337 @@
+// TODO: Rename to backup.mjs before committing
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backup.mjs]: Done in commit 801f72307.

File: backup.mjs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+console.warn(`Fetching project with ID ${projId} from ${contextName} context, named "${context}"`);
+console.warn("If that looks wrong, hit Ctrl+C right NOW!");
+console.warn();
+console.warn("Pausing for 2 seconds to give you time to hit Ctrl+C...");
+await new Promise((resolve) => setTimeout(resolve, 2000));
+// Start running
+
+console.warn("Setting up kubectl port forwarding for remote Mongo...");
+const remoteMongoPort = await randomFreePort();
+// TODO: Improve by finding a local port that's not in use, rather than hardcoding this
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backup.mjs]: Done in commit 801f72307.

**Comment by hahn-kev** [comment]
I want to say windows typically uses double quote's (`"`) instead of single to do something like that.

**Comment by rmunn** [comment]
> I want to say windows typically uses double quote's (`"`) instead of single to do something like that.

That makes things difficult, because Linux assigns different meaning to single quotes vs double quotes; for example, single quotes don't do `$variable` expansion. There are other differences too, so I use single-quotes routinely when doing `sh -c 'some long command'` because that works far more often. Plus it allows me to put double-quotes inside the command.

I might be able to rewrite the `exec` calls to use the version of `exec` where you pass each parameter as a separate string and let Node take care of appropriate quoting on each OS.

**Comment by megahirt** [comment]
> That makes things difficult, because Linux assigns different meaning to single quotes vs double quotes;

Yup, that's true.  Are you saying you cannot make it work with double quotes?
```
`kubectl --context=${context} --namespace=languageforge exec -c app deploy/app -- sh -c "readlink -eq ${name} >/dev/null && echo yes || echo no"`,
 ```
Does it work on Linux like this?

**Comment by rmunn** [comment]
> ```
> `kubectl --context=${context} --namespace=languageforge exec -c app deploy/app -- sh -c "readlink -eq ${name} >/dev/null && echo yes || echo no"`,
> ```
> 
> Does it work on Linux like this?

Just pushed a commit making it work with double quotes, which I started even before your comment. :-) I would have pushed it an hour ago, but DockerHub was giving me just a trickle of bandwidth so it took nearly an hour to download the Docker images and run local LF to test it.

**Comment by rmunn** [comment]
Just found a bug: if the lexicon collection exists but is empty, you get `MongoInvalidArgumentError: Invalid BulkOperation, Batch cannot be empty`.

I'll fix it, but if it takes too long then I won't spend too much time on it — because a project with no lexical entries at all is not one that we're likely to need to copy to local LF in order to troubleshoot. :-)

**Review by myieye — COMMENTED** [review]
Unfortunately, I'm pretty stuck.
The copying the assets tar to my Windows machine has multiple problems.
1) I had to normalize the path (actually, I'm not sure if that was necessary, because it didn't help)
2) It seems that kubectl cp interprets the Drive in a windows path as a pod name (after the bug has already been fixed twice, apparently 😉). https://github.com/kubernetes/kubernetes/issues/101985#issuecomment-2143893126
3) I can't use a relative path, because my code is on D and my temp folder on C
4) When I copied the code to C to try it out I got:
```
error: unexpected EOF
kubectl cp failed Error: Command failed: kubectl --context="aws-rke" --namespace=languageforge cp app-7659cf6f57-n9cmx:/tmp/assets-sf_test-chris-02_lf.tar ..\..\AppData\Local\Temp\lfbackup-i56u10\assets-sf_test-chris-02_lf.tar 
error: unexpected EOF
. Will try to continue with rsync...
Ensuring rsync exists in target container...
Error from server (Forbidden): deployments.apps "app" is forbidden: User "u-j7l8z" cannot get resource "deployments" in API group "apps" in the namespace "default"
Cleaning up temporary directory C:\Users\tim\AppData\Local\Temp\lfbackup-i56u10...
node:internal/errors:932
  const err = new Error(message);
              ^

Error: Command failed: kubectl exec --context="aws-rke" -c app deploy/app -- bash -c "which rsync || (apt update && apt install rsync -y)"
Error from server (Forbidden): deployments.apps "app" is forbidden: User "u-j7l8z" cannot get resource "deployments" in API group "apps" in the namespace "default"
```

I have no idea what the `error: unexpected EOF` is about 🙁.

I don't really want to set up k8s in WSL just to test this 😕.

If it's working now in Linux, then maybe you should just merge it. 🤷

**Comment by rmunn** [comment]
The "unexpected EOF" error is because `kubectl cp` is unreliable, and will continue to be unreliable until the server is running Kubernetes 1.30 or later *and* you have a kubectl with version 1.30 or later. That's precisely why I designed the script to fall back to rsync.

The bit about `deploy/app` not being available to your user account is fixable: I already have the pod name from the earlier kubectl cp step, so I should use use the pod name instead of `deploy/app`. So I'll fix that first.

> If it's working now in Linux, then maybe you should just merge it. 🤷

Okay, I'll dismiss your "changes requested" review from earlier so that GitHub will allow me to merge this.

**Comment by rmunn** [comment]
You might be able to work around the "drive letter interpreted as a pod name" error by using the obscure `\\localhost\c$\my_dir` format for paths. https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#unc-paths

**Comment by rmunn** [comment]
Feature request from meeting: we want to auto-cleanup the tar file from the server on script exit, so we don't leave a bunch of asset tarballs lying around until the next container restart.

**Comment by rmunn** [comment]
@myieye -

As we discussed, leaving this bit of the work for you. Commit https://github.com/sillsdev/web-languageforge/pull/1816/commits/45d82949f933077d72ce0bec8e2d61200334fd76 adds a comment in the place where you'd want to make that substitution.

> You might be able to work around the "drive letter interpreted as a pod name" error by using the obscure `\\localhost\c$\my_dir` format for paths. https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#unc-paths

**Review by myieye — APPROVED** [review]
I made the path change you suggested ([1e3ced2](https://github.com/sillsdev/web-languageforge/pull/1816/commits/1e3ced21865a8614e747599303306c3cd00d2a38)) and it worked! 🥳 
image

File: backup.mjs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply backup.mjs]: Done now.

**Pull request merged**